library/alloc/src/boxed.rs

```diff
@@ -1016,6 +1016,177 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
+        if Box::try_set(b, val).is_err() {
+            handle_alloc_error(layout);
+        }
+    }
+
+    /// Sets the value inside the box, reusing the existing storage if possible and reallocating if
+    /// not, returning an error if reallocation failed.
+    ///
+    /// If the old and new values stored in the box have the same [memory layout](Layout), then the
+    /// contents of the `Box` will be updated in-place and no memory allocation will occur.
+    /// Otherwise, a new `Box` will be created to replace the old one, causing a memory allocation.
+    /// To instead return an error if this happens, use [`Box::reuse`].
+    ///
+    /// # Examples
+    ///
+    /// Reuse a boxed slice:
+    ///
+    /// ```
+    /// #![feature(allocator_api, box_reuse)]
+    ///
+    /// let mut numbers: Box<[i32]> = Box::try_new([1, 2, 3])?;
+    ///
+    /// // This will not reallocate.
+    /// Box::try_set(&mut numbers, [4, 5, 6])?;
+    /// assert_eq!(&*numbers, &[4, 5, 6]);
+    ///
+    /// // This will reallocate.
+    /// Box::try_set(&mut numbers, [4, 5, 6, 7])?;
+    /// assert_eq!(&*numbers, &[4, 5, 6, 7]);
+    /// # Ok::<(), std::alloc::AllocError>(())
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    // #[unstable(feature = "allocator_api", issue = "32838")]
+    pub fn try_set<U>(b: &mut Box<T, A>, val: U) -> Result<(), AllocError>
+    where
+        U: Unsize<T>,
+    {
+        if let Err(val) = Box::reuse(b, val) {
+            // Due to https://github.com/rust-lang/rust/issues/78459, it is not currently possible
+            // to use a `Box<T, &A>`, so we have to work around it.
+
+            // Allocate the replacement box now so we don't have to deal with code panicking later.
+            let alloc = Box::allocator(b);
+            // Use allocation methods directly instead of `Box::try_new_in` so ownership of the
+            // allocator is not lost if allocation errors or panics.
+            let ptr: *mut U = alloc.allocate(Layout::new::<U>())?.cast().as_ptr();
+            unsafe { ptr::write(ptr, val) };
+            let new_b = unsafe { Box::from_raw_in(ptr, ptr::read(alloc)) };
+
+            let old_b = mem::replace(b, new_b);
+
+            // Drop the old box without dropping its allocator, as the allocator has been moved
+            // into the new box.
+            struct DeallocGuard<A: Allocator> {
+                alloc: A,
+                ptr: ptr::NonNull<u8>,
+                layout: Layout,
+            }
+            impl<A: Allocator> Drop for DeallocGuard<A> {
+                fn drop(&mut self) {
+                    unsafe { self.alloc.deallocate(self.ptr, self.layout) };
+                }
+            }
+
+            let old_val = Box::leak(old_b);
+            let _dealloc_guard = DeallocGuard {
+                alloc: Box::allocator(&b),
+                ptr: ptr::NonNull::from(&mut *old_val).cast(),
+                layout: Layout::for_value(old_val),
```

Comment:
```suggestion
                layout: Layout::for_value(old_val),
                ptr: ptr::NonNull::from(old_val).cast(),
```

Reply:
This wouldn't compile because it consumes `old_val`, but that reference needs to be used later for the `ptr::drop_in_place`.

Reply:
Fair enough, although the later code could use `guard.ptr` to perform the drop_in_place. The advantage of this approach is that it is more obvious to me that full ownership of `old_val` was passed to the guard, and then it is up to our usage of `guard` to see how it handles both the pointee and its backing allocation

Reply:
Thanks for the advice. I like that approach, I have implemented it.